### PR TITLE
onnx rnn export: use spec-respecting dimensions

### LINKED ISSUE #6879: [ONNX] LSTM weight tensors: wrong dimensions
## Issue description

The [ONNX specification](https://github.com/onnx/onnx/blob/master/docs/Operators.md#inputs-3---8) requires the following dimensions for the weight tensors `W`, `R`, and `B` of an LSTM:

`W`: `[num_directions, 4*hidden_size, input_size]` (rank 3)
`R`: `[num_directions, 4*hidden_size, hidden_size]` (rank 3)
`B`: `[num_directions, 8*hidden_size]` (rank 2)

However, when exporting an LSTM from PyTorch, the tensors used for the LSTM weights in the resulting model have the following dimensions:

`W`: `[num_directions*4*hidden_size, input_size]` (rank 2)
`R`: `[num_directions*4*hidden_size, hidden_size]` (rank 2)
`B`: `[num_directions*8*hidden_size]` (rank 1),

i.e. the weights for the different directions are concatenated, instead of being stored in separate dimensions. This does not conform to the specification.

## Code example

```python
import torch

inputs = torch.Tensor(5,1,3)
lstm = torch.nn.LSTM(3, 3, bidirectional=True) # Input dim = hidden dim = 3
out, hidden = lstm(inputs)  

torch.onnx.export(lstm, inputs, "lstm.onnx")
```

According to the specification, the weight tensors of the LSTM should have the following dimensions:

`W`: `[2, 12, 3]` (rank 3)
`R`: `[2, 12, 3]` (rank 3)
`B`: `[2, 24]` (rank 2)

However, inspecting `lstm.onnx` yields the following dimensions:

`W`: `[24, 3]` (rank 2)
`R`: `[24, 3]` (rank 2)
`B`: `[48]` (rank 1).

### PR DESCRIPTION
**Pull request by anderspapitto**
fixes https://github.com/pytorch/pytorch/issues/6879

### CODE REVIEW

File: torch/onnx/symbolic.py [hunk 1]
```diff
@@ -911,9 +941,20 @@ def symbolic(g, input, all_weights, h0, batch_sizes):
...
+                weight_hh = g.op('Concat',
+                                 g.op('Unsqueeze', weight_hh_f, axes_i=[0]),
+                                 g.op('Unsqueeze', weight_hh_b, axes_i=[0]),
+                                 axis_i=0)
+                bias_f = g.op('Concat', bias_ih_f, bias_hh_f, axis_i=0)
+                bias_b = g.op('Concat', bias_ih_b, bias_hh_b, axis_i=0)
+                bias_concat = g.op('Concat',
+                                   g.op('Unsqueeze', bias_f, axes_i=[0]),
+                                   g.op('Unsqueeze', bias_b, axes_i=[0]),
+                                   axis_i=0)
```
Comment (ezyang) [inline torch/onnx/symbolic.py]: The diff looks fairly repetitive, is there really no duplication going on here?

File: torch/onnx/symbolic.py [hunk 2]
```diff
@@ -795,16 +795,31 @@ def symbolic(g, input, all_weights, h0, batch_sizes):
             if unidirectional:
                 weight_ih, weight_hh, bias_ih, bias_hh = all_weights[i]
                 bias_concat = g.op('Concat', bias_ih, bias_hh, axis_i=0)
```
Comment (ezyang) [inline torch/onnx/symbolic.py]: Maybe add a comment like:

```
# Unidirectional weights in PyTorch omit the directionality dimension (dim=0), whereas in ONNX
# it is required even for unidirectional RNN; thus, insert a dimension of size at dim=0.
```

**Review by ezyang — APPROVED** [review]

File: torch/onnx/symbolic.py (hunk 1 above)
Comment (anderspapitto) [reply torch/onnx/symbolic.py]: There is duplication, which I plan to fix but not in this diff. I recently fixed the analogous code duplication on the onnx-caffe2 side here https://github.com/pytorch/pytorch/pull/7263

File: torch/onnx/symbolic.py (hunk 2 above)
Comment (anderspapitto) [reply torch/onnx/symbolic.py]: Ok, will add this comment

**Comment by anderspapitto** [comment]
complimented by https://github.com/onnxbot/onnx-fb-universe/pull/2012

looks like there are also a couple other test failures which I am investigating

**Comment by anderspapitto** [comment]
https://github.com/onnx/onnx/pull/926 regenerates files used by the failing caffe2 tests

**Pull request merged**